### PR TITLE
Add V2 Light System for Old Environment

### PR DESCRIPTION
**Pull request by CCDV2**
# Add V2 Light System for Old Environments
Introducing v2 light system, including:
- White Lights (default keymap is `3`, which is a same reference to bomb)
- Transition event (integer value 4, 8, 12 under light event type, and default keymap is `c`)
- V3 NOT environment supported. (Weave, FOB, EDM)

## Existing Problems
- Lighting may not be similar as official editor or in-game (of course...)
- No guarantee of fade/flash together with transition.

### CODE REVIEW

**Comment by XAce1337manX** [comment]
Other comments:
* Transitions don't preview increasing alpha for some reason. It does for decreasing.
* Transitions currently don't react to `Middle Mouse Click` unlike the other existing types
* Transition preview logic breaks on shifting events to other lanes

**Comment by CCDV2** [comment]
Noticed. Will fix them later.

File: Assets/__Scripts/Platforms/PlatformDescriptor.cs [hunk 1]
```diff
@@ -382,13 +387,70 @@ private void HandleLights(LightsManager group, int value, MapEvent e)
...
+    private void TrySetTransition(LightingEvent light, MapEvent e)
+    {
+        if (TryGetNextTransitionNote(e, out var transition))
+        {
+            var targetAlpha = transition.FloatValue;
+            var transitionTime = atsc.GetSecondsFromBeat(transition.Time - e.Time);
+            var targetColor = InferColorFromValue(light.UseInvertedPlatformColors, transition.Value);
+            light.UpdateTargetColor(targetColor.Multiply(LightsManager.HDRIntensity), transitionTime);
+            light.UpdateTargetAlpha(targetAlpha, transitionTime);
+            light.UpdateEasing(Mathf.Approximately(targetAlpha, 0) ? "easeOutExpo" : "easeLinear"); // may not be the correct easing function
```
[the remark this answers is not in the document]
Comment (CCDV2) [reply Assets/__Scripts/Platforms/PlatformDescriptor.cs:450]: For strobe light, linear transition looks good. For fading to zero, linear transition will make a sudden luminance drop(I think). I don't know how to better unify or describe them.

**Comment by CCDV2** [comment]
I commented `UpdateMultiplyAlpha` so that increasing alpha transition could work. However I don't know if it breaks some other places or not..

**Comment by brochasep** [comment]
You need to make sure it can be used in conjunction with v3 since the environments have some v2 things, even when they use v3 lighting.

**Comment by CCDV2** [comment]
A few updates:
- IsControlLight is derived from IsUtilityEvent now.
- Two modifier for simply put 0.00/0.50 float value light for v2 light system. Default keybinds are `z` and `x`, respectively(I can't  use shift/ctrl/alt because there are existing keybindings for them + click).

**Comment by XAce1337manX** [comment]
Default keybind for `x` conflicts with default changing precision.

Transition functionality seems ok.

**Comment by CCDV2** [comment]
> Default keybind for `x` conflicts with default changing precision.
> 
> Transition functionality seems ok.

I cannot find other empty keybind.. So I'm currently reusing `x`..

**Comment by CCDV2** [comment]
A Small Update:
I don't know whether `color.a` channel is used for some places. Previous condition only consider floatvalue with 1 so it will have bug with boost. So I commented one line.

**Pull request merged**